File: src/Vograph.Core/Services/OverrideService.cs
using Vograph.Core.Models;
using Vograph.Core.Services.Sync;

namespace Vograph.Core.Services;

public class OverrideService
{
    private readonly Database _db;
    private readonly PrivateSyncOutbox? _outbox;
    public OverrideService(Database db, PrivateSyncOutbox? outbox = null)
    {
        _db = db;
        _outbox = outbox;
    }

    public string GetDisplayName(string subjectRaw, int dayOfWeek)
    {
        var norm = ParityService.NormalizeSubject(subjectRaw);
        var overrides = _db.GetOverrides();
        // global overrides weekday
        var global = overrides.FirstOrDefault(o => o.SubjectRawNormalized == norm && o.Scope == "global");
        if (global != null) return global.DisplayName;
        var weekday = overrides.FirstOrDefault(o => o.SubjectRawNormalized == norm && o.Scope == $"weekday:{dayOfWeek}");
        if (weekday != null) return weekday.DisplayName;
        return subjectRaw;
    }

    public Override? GetOverride(string subjectRaw, string scope)
    {
        var norm = ParityService.NormalizeSubject(subjectRaw);
        return _db.GetOverrides().FirstOrDefault(o => o.SubjectRawNormalized == norm && o.Scope == scope);
    }

    public long AddOrUpdate(string subjectRaw, string scope, string displayName, string? note)
    {
        if (_outbox is { Enabled: true })
            return _outbox.InTransaction(() => AddOrUpdateCore(subjectRaw, scope, displayName, note, enqueue: true));
        return AddOrUpdateCore(subjectRaw, scope, displayName, note, enqueue: false);
    }

    private long AddOrUpdateCore(string subjectRaw, string scope, string displayName, string? note, bool enqueue)
    {
        var norm = ParityService.NormalizeSubject(subjectRaw);
        var existing = GetOverride(subjectRaw, scope);
        if (enqueue && existing != null)
        {
            existing.EntityId ??= Guid.NewGuid();
            existing.CreatedAtUtc ??= DateTimeOffset.UtcNow;
            using var upd = _db.Connection.CreateCommand();
            upd.CommandText = "UPDATE overrides SET displayName=@d, note=@n, entityUuid=@u, createdAtUtc=@utc WHERE id=@id";
            upd.Parameters.AddWithValue("@d", displayName);
            upd.Parameters.AddWithValue("@n", (object?)note ?? DBNull.Value);
            upd.Parameters.AddWithValue("@u", existing.EntityId.Value.ToString("D"));
            upd.Parameters.AddWithValue("@utc", existing.CreatedAtUtc.Value.ToString("o"));
            upd.Parameters.AddWithValue("@id", existing.Id);
            upd.ExecuteNonQuery();
            var value = PrivateSyncMapper.Override(norm, scope, displayName, note, existing.CreatedAtUtc.Value);
            _outbox!.Enqueue(Guid.NewGuid(), "override", existing.EntityId.Value, existing.Revision, "upsert", value, existing.Id);
            return existing.Id;
        }
        if (existing != null)
        {
            using var cmd = _db.Connection.CreateCommand();
            cmd.CommandText = "DELETE FROM overrides WHERE id=@id";
            cmd.Parameters.AddWithValue("@id", existing.Id);
            cmd.ExecuteNonQuery();
        }
        var created = DateTime.UtcNow;
        var utc = new DateTimeOffset(created, TimeSpan.Zero);
        var ov = new Override
        {
            SubjectRawNormalized = norm,
            Scope = scope,
            DisplayName = displayName,
            Note = note,
            CreatedAt = created,
            EntityId = enqueue ? Guid.NewGuid() : null,
            CreatedAtUtc = enqueue ? utc : null
        };
        var id = _db.InsertOverride(ov);
        if (enqueue)
            _outbox!.Enqueue(Guid.NewGuid(), "override", ov.EntityId!.Value, 0, "upsert",
                PrivateSyncMapper.Override(norm, scope, displayName, note, utc), id);
        return id;
    }

    public void Remove(long id)
    {
        if (_outbox is { Enabled: true })
        {
            _outbox.InTransaction(() =>
            {
                var existing = _db.GetOverrides().FirstOrDefault(o => o.Id == id);
                if (existing is null) return 0;
                existing.EntityId ??= Guid.NewGuid();
                if (existing.Revision == 0)
                {
                    using var hard = _db.Connection.CreateCommand();
                    hard.CommandText = "DELETE FROM overrides WHERE id=@id";
                    hard.Parameters.AddWithValue("@id", id);
                    hard.ExecuteNonQuery();
                    _outbox.Enqueue(Guid.NewGuid(), "override", existing.EntityId.Value, 0, "delete", null, id);
                    return 0;
                }
                using var cmd = _db.Connection.CreateCommand();
                cmd.CommandText = "UPDATE overrides SET tombstone=1, entityUuid=@u WHERE id=@id";
                cmd.Parameters.AddWithValue("@u", existing.EntityId.Value.ToString("D"));
                cmd.Parameters.AddWithValue("@id", id);
                cmd.ExecuteNonQuery();
                _outbox.Enqueue(Guid.NewGuid(), "override", existing.EntityId.Value, existing.Revision, "delete", null, id);
                return 0;
            });
            return;
        }
        using var legacy = _db.Connection.CreateCommand();
        legacy.CommandText = "DELETE FROM overrides WHERE id=@id";
        legacy.Parameters.AddWithValue("@id", id);
        legacy.ExecuteNonQuery();
    }

    public string? GetNote(string subjectRaw, int dayOfWeek)
    {
        var norm = ParityService.NormalizeSubject(subjectRaw);
        var overrides = _db.GetOverrides();
        var global = overrides.FirstOrDefault(o => o.SubjectRawNormalized == norm && o.Scope == "global");
        if (global != null) return global.Note;
        var weekday = overrides.FirstOrDefault(o => o.SubjectRawNormalized == norm && o.Scope == $"weekday:{dayOfWeek}");
        return weekday?.Note;
    }
}

File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/OverrideService.kt
package ru.bgtu_voenmeh.zapara.data

import ru.bgtu_voenmeh.zapara.data.db.OverrideDao
import ru.bgtu_voenmeh.zapara.data.db.OverrideEntity
import ru.bgtu_voenmeh.zapara.data.sync.OverrideValue
import ru.bgtu_voenmeh.zapara.data.sync.RoomSyncOutbox
import ru.bgtu_voenmeh.zapara.data.sync.SyncLocalIdentity
import ru.bgtu_voenmeh.zapara.data.sync.SyncValidation
import java.time.LocalDate
import java.util.UUID

// Port of Vograph.Core OverrideService. Global scope wins over weekday scope.
class OverrideService(
    private val dao: OverrideDao,
    private val outbox: RoomSyncOutbox? = null
) {

    fun addOrUpdate(subjectRaw: String, scope: String, displayName: String, note: String?): Long {
        if (outbox?.enabled == true) {
            return outbox.inTransaction { addOrUpdateCore(subjectRaw, scope, displayName, note, enqueue = true) }
        }
        return addOrUpdateCore(subjectRaw, scope, displayName, note, enqueue = false)
    }

    private fun addOrUpdateCore(
        subjectRaw: String,
        scope: String,
        displayName: String,
        note: String?,
        enqueue: Boolean
    ): Long {
        val norm = Parity.normalizeSubject(subjectRaw)
        val existing = dao.getAll().firstOrNull { it.subjectRawNormalized == norm && it.scope == scope }
        if (enqueue && existing != null) {
            val ident = ensureOverrideIdentity(existing)
            dao.deleteById(existing.id)
            dao.insert(existing.copy(displayName = displayName, note = note, id = existing.id))
            val value = OverrideValue(
                norm,
                SyncValidation.normalizeSubject(norm),
                scope,
                displayName,
                note,
                ident.createdAtUtc
            )
            outbox!!.enqueue(UUID.randomUUID(), "override", ident.entityId, ident.revision, "upsert", value, existing.id)
            return existing.id
        }
        if (existing != null) {
            dao.deleteByKey(norm, scope)
        }
        val id = dao.insert(
            OverrideEntity(
                subjectRawNormalized = norm, scope = scope,
                displayName = displayName, note = note,
                createdAt = LocalDate.now().toString()
            )
        )
        if (enqueue) {
            val raw = if (subjectRaw.isBlank()) subjectRaw else subjectRaw.trim()
            val utc = outbox!!.nowUtc()
            val entityId = UUID.randomUUID()
            outbox.remember("override", id, entityId, 0, utc, null)
            outbox.enqueue(
                UUID.randomUUID(),
                "override",
                entityId,
                0,
                "upsert",
                OverrideValue(raw, SyncValidation.normalizeSubject(raw), scope, displayName, note, utc),
                id
            )
        }
        return id
    }

    fun displayName(subjectRaw: String, dayOfWeek: Int): String {
        val found = displayNameByNorm(Parity.normalizeSubject(subjectRaw), dayOfWeek)
        return found.ifEmpty { subjectRaw }
    }

    fun displayNameByNorm(norm: String, dayOfWeek: Int): String {
        val all = dao.getAll().filter { it.subjectRawNormalized == norm }
        all.firstOrNull { it.scope == "global" }?.let { return it.displayName }
        all.firstOrNull { it.scope == "weekday:$dayOfWeek" }?.let { return it.displayName }
        return ""
    }

    fun note(subjectRaw: String, dayOfWeek: Int): String {
        return noteByNorm(Parity.normalizeSubject(subjectRaw), dayOfWeek)
    }

    fun noteByNorm(norm: String, dayOfWeek: Int): String {
        val all = dao.getAll().filter { it.subjectRawNormalized == norm }
        val ov = all.firstOrNull { it.scope == "global" }
            ?: all.firstOrNull { it.scope == "weekday:$dayOfWeek" }
        return ov?.note.orEmpty()
    }

    fun remove(id: Long) {
        if (outbox?.enabled == true) {
            outbox.inTransaction {
                val existing = dao.getAll().firstOrNull { it.id == id } ?: return@inTransaction 0
                val ident = ensureOverrideIdentity(existing)
                dao.deleteById(id)
                outbox.enqueue(UUID.randomUUID(), "override", ident.entityId, ident.revision, "delete", null, id)
                0
            }
            return
        }
        dao.deleteById(id)
    }

    private fun ensureOverrideIdentity(existing: OverrideEntity): SyncLocalIdentity {
        outbox!!.identity("override", existing.id)?.let { return it }
        val utc = outbox.nowUtc()
        return outbox.remember("override", existing.id, UUID.randomUUID(), 0, utc, null)
    }

    fun all(): List<OverrideEntity> = dao.getAll()
}
